### PR TITLE
add pre-built detection

### LINKED ISSUE #81468: Enable source-build pre-built detection
Part of https://github.com/dotnet/source-build/issues/3017

Enable source-build pre-build detection on the current repository and resolve any pre-build issues discovered

### PR DESCRIPTION
**Pull request by oleksandr-didyk**
Resolves https://github.com/dotnet/runtime/issues/81468

Existing issues:

- [x] `Microsoft.CodeAnalysis 4.5.0` - https://github.com/dotnet/source-build-reference-packages/pull/669
- [x] `DotNet.Build.Tasks` issue - https://github.com/dotnet/runtime/pull/86151#issuecomment-1549669873, solved by https://github.com/dotnet/source-build-reference-packages/pull/673
- [x] `runtime/src/libraries/pretest.proj` - source-build logic has a [check](https://github.com/dotnet/arcade/blob/f38f6d1439ff6546a97f568f3e6bd3eea3ac8e6c/src/Microsoft.DotNet.SourceBuild/tasks/src/UsageReport/WriteUsageReports.cs#L253) for test projects & the check complains about `src/libraries/pretest.proj`. Silenced it for the repo as requested in https://github.com/dotnet/runtime/pull/86151#issuecomment-1548252281. Additionally, created https://github.com/dotnet/source-build/issues/3453 to get rid of the check completely

CC: @MichaelSimons @ViktorHofer

### CODE REVIEW

**Comment by ViktorHofer** [comment]
> starting from Arcade version 8.0.0-beta.23260.3 DotNet.Build.Tasks.TargetFrameworks.targets tries to load Nuget.Frameworks from SBRP during the build and fails

We have been relying on NuGet APIs underneath the Microsoft.DotNet.Build.Tasks.TargetFramework package for years. What's the reason for this showing up now?

**Comment by oleksandr-didyk** [comment]
> > starting from Arcade version 8.0.0-beta.23260.3 DotNet.Build.Tasks.TargetFrameworks.targets tries to load Nuget.Frameworks from SBRP during the build and fails
> 
> We have been relying on NuGet APIs underneath the Microsoft.DotNet.Build.Tasks.TargetFramework package for years. What's the reason for this showing up now?

I added SRBP for it as part of this work, which was fine at first, but after my change in Arcade the build started failing as it was loading the ref pack in

**Comment by ViktorHofer** [comment]
> source-build logic has a [check](https://github.com/dotnet/arcade/blob/f38f6d1439ff6546a97f568f3e6bd3eea3ac8e6c/src/Microsoft.DotNet.SourceBuild/tasks/src/UsageReport/WriteUsageReports.cs#L253) for test projects & the check complains about src/libraries/pretest.proj. I can find a way to silence the check for this particular case. That, or maybe the proj should not have test in its name in the first place?

Yes, please silence the check. I'm not exactly happy about that validation tbh. We have other shipping projects with "Tests" in its name which currently are excluded from source build as they rely on external prebuilds, i.e. Microsoft.Extensions.DependencyInjection.Specifications.Tests. While those currently don't cause a problem, we might have other such cases in the future which we want to include during source build.

**Comment by oleksandr-didyk** [comment]
> DotNet.Build.Tasks issue - starting from Arcade version 8.0.0-beta.23260.3 DotNet.Build.Tasks.TargetFrameworks.targets tries to load Nuget.Frameworks from SBRP during the build and fails ([source](https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.Build.Tasks.TargetFramework/src/build/Microsoft.DotNet.Build.Tasks.TargetFramework.targets?rgh-link-date=2023-05-12T14%3A07%3A26Z#L55)); the version where this manifests has only one change, which is bump of DependencyModel to latest, i.e. I probably missed something during testing. Will investigate.

I think I figured out what happened, but will talk it out with Michael to be sure -> `DotNet.Build.Tasks.TargetFrameworks` package comes with a version of `Nuget` already inside of it, in the `tools` directory. This version is the one that gets loaded in.

In this case, it seems that during Arcade build the SBRP version of Nuget got included in the `tools` directory and gets loaded in, crashing the build. I verified this building the latest Arcade version without SBRP and then using that version in `runtime`

File: eng/Versions.props [hunk 1]
```diff
@@ -93,7 +93,9 @@
     <MicrosoftDotNetBuildTasksArchivesVersion>8.0.0-beta.23265.1</MicrosoftDotNetBuildTasksArchivesVersion>
     <MicrosoftDotNetBuildTasksInstallersVersion>8.0.0-beta.23265.1</MicrosoftDotNetBuildTasksInstallersVersion>
     <MicrosoftDotNetBuildTasksPackagingVersion>8.0.0-beta.23265.1</MicrosoftDotNetBuildTasksPackagingVersion>
-    <MicrosoftDotNetBuildTasksTargetFrameworkVersion>8.0.0-beta.23265.1</MicrosoftDotNetBuildTasksTargetFrameworkVersion>
+    <!-- Will be removed once latest SBRP flows into Arcade -->
```
Comment (MichaelSimons) [inline eng/Versions.props:96]: You are adding the SBRP dependency - Can't you update the reference now?

Comment (oleksandr-didyk) [reply eng/Versions.props:96]: I am just building the repo locally first to verify the change, will push the fix for it right after

Comment (oleksandr-didyk) [reply eng/Versions.props]: Removed

File: eng/SourceBuildPrebuiltBaseline.xml [hunk 2]
```diff
@@ -1,5 +1,22 @@
...
+    <!-- Allowed and pinned to major version due to https://github.com/dotnet/source-build/issues/3228 -->
+    <UsagePattern IdentityGlob="Microsoft.NETCore.App.Crossgen2.linux-x64/*8.*" />
+    <UsagePattern IdentityGlob="Microsoft.NETCore.App.Runtime.linux-x64/*8.*" />
+    <UsagePattern IdentityGlob="*Microsoft.DotNet.ILCompiler/*8.*" />
+    
+    <UsagePattern IdentityGlob="Microsoft.CodeAnalysis.Analyzers/*" />
   </IgnorePatterns>
-</UsageData>
+  <Usages>
+    <!-- Should be removed as part of https://github.com/dotnet/runtime/issues/82298 -->
```
[the remark this answers is not in the document]
Comment (oleksandr-didyk) [reply eng/SourceBuildPrebuiltBaseline.xml]: Thanks for pointing it out, removed

**Review by MichaelSimons — APPROVED** [review]

**Review by ViktorHofer — APPROVED** [review]

**Comment by lewing** [comment]
in the future please make sure to update the shas too when updating VersionDetails.xml

**Comment by oleksandr-didyk** [comment]
> in the future please make sure to update the shas too when updating VersionDetails.xml

My bad, normally I update dependencies via the darc client, which handles this for me, but this time I did it manually. Will make sure to not forget this in the future

**Pull request merged**